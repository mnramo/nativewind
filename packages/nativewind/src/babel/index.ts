import { resolve, sep, posix } from "node:path";
import { statSync } from "node:fs";

import micromatch from "micromatch";

import type { Config } from "tailwindcss";
import resolveConfigPath from "tailwindcss/lib/util/resolveConfigPath";
import resolveConfig from "tailwindcss/resolveConfig";
import { validateConfig } from "tailwindcss/lib/util/validateConfig";

import type { ConfigAPI, PluginPass, Visitor } from "@babel/core";
import { addNamed } from "@babel/helper-module-imports";

import { extractStyles } from "../postcss/extract-styles";
import { TailwindcssReactNativeBabelOptions } from "./types";
import { visitor, VisitorState } from "./visitor";

import { nativePlugin } from "../tailwind/native";
import { expressionStatement } from "@babel/types";

export default function (
  api: ConfigAPI,
  options: TailwindcssReactNativeBabelOptions,
  cwd: string
) {
  /**
   * Get the users config
   */
  const userConfigPath = resolveConfigPath(
    options.tailwindConfig || options.tailwindConfigPath
  );

  let tailwindConfig: Config;

  if (userConfigPath === null) {
    tailwindConfig = resolveConfig({
      ...options.tailwindConfig,
      plugins: [
        nativePlugin(options),
        ...(options?.tailwindConfig?.plugins ?? []),
      ],
    });
  } else {
    api.cache.using(() => statSync(userConfigPath).mtimeMs);
    // eslint-disable-next-line  @typescript-eslint/no-explicit-any
    (api as any).addExternalDependency(userConfigPath);

    // eslint-disable-next-line @typescript-eslint/no-var-requires,unicorn/prefer-module
    const userConfig = require(userConfigPath);

    const newConfig = resolveConfig({
      ...userConfig,
      plugins: [nativePlugin(options), ...(userConfig?.plugins ?? [])],
    });

    tailwindConfig = validateConfig(newConfig);
  }

  /**
   * Resolve their content paths
   */
  const contentFilePaths = (
    Array.isArray(tailwindConfig.content)
      ? tailwindConfig.content.filter(
          (filePath): filePath is string => typeof filePath === "string"
        )
      : tailwindConfig.content.files.filter(
          (filePath): filePath is string => typeof filePath === "string"
        )
  ).map((contentFilePath) => normalizePath(resolve(cwd, contentFilePath)));

  const allowModuleTransform = Array.isArray(options.allowModuleTransform)
    ? ["react-native", "react-native-web", ...options.allowModuleTransform]
    : "*";

  const programVisitor: Visitor<
    PluginPass & {
      opts: TailwindcssReactNativeBabelOptions;
    }
  > = {
    Program: {
      enter(projectPath, state) {
        const filename = state.filename;

        if (!filename) return;

        const isInContent = micromatch.isMatch(
          normalizePath(filename),
          contentFilePaths
        );

        if (!isInContent) {
          return;
        }

        let canCompile = true;
        let canTransform = true;

        switch (state.opts.mode) {
          case "compileOnly": {
            canTransform = false;
            break;
          }
          case "transformOnly": {
            canCompile = false;
            break;
          }
        }

        const visitorState: VisitorState = {
          ...state,
          rem: 16,
          didTransform: false,
          canCompile,
          canTransform,
          filename,
          allowModuleTransform,
          allowRelativeModules: contentFilePaths,
          blockList: new Set(),
          tailwindConfig: tailwindConfig,
        };

        projectPath.traverse(visitor, visitorState);

        const { didTransform } = visitorState;

        if (didTransform) {
          addNamed(projectPath, "StyledComponent", "nativewind");
        }

        const output = extractStyles({
          ...tailwindConfig,
          content: [filename],
          // If the file doesn't have any Tailwind styles, it will print a warning
          // We force an empty style to prevent this
          safelist: tailwindConfig.safelist ?? ["babel-empty"],
        });

        if (!output.hasStyles) return;

        projectPath.pushContainer(
          "body",
          expressionStatement(output.stylesheetCreateExpression)
        );

        addNamed(projectPath, "NativeWindStyleSheet", "nativewind");
      },
    },
  };

  return {
    visitor: programVisitor,
  };
}

function normalizePath(filePath: string) {
  /**
   * This is my naive way to get path matching working on Windows.
   * Basically I turn it into a posix path which seems to work fine
   *
   * If you are a windows user and understand micromatch, can you please send a PR
   * to do this the proper way
   */
  return filePath.split(sep).join(posix.sep);
}
