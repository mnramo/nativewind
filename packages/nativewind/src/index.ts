export * from "./styled";
export * from "./styled-component";
export * from "./styled/use-tailwind";
export { NativeWindStyleSheet } from "./style-sheet";
export { useColorScheme } from "./use-color-scheme";
export { platformSelect } from "./utils/platform-select";
export { withExpoSnack } from "./expo-snack";
export type { StyledProps } from "./utils/styled";
