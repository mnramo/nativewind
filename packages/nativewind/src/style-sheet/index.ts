import { createContext } from "react";
import { StyleSheetRuntime } from "./runtime";

export type { StylesArray, Snapshot } from "./runtime";
export { StyleSheetRuntime } from "./runtime";

const runtime = new StyleSheetRuntime();

export const NativeWindStyleSheet = {
  create: runtime.create.bind(runtime),
  setDimensions: runtime.setDimensions.bind(runtime),
  setAppearance: runtime.setAppearance.bind(runtime),
  setPlatform: runtime.setPlatform.bind(runtime),
  setOutput: runtime.setOutput.bind(runtime),
  setColorScheme: runtime.setColorScheme.bind(runtime),
  parse: runtime.parse.bind(runtime),
  setDangerouslyCompileStyles:
    runtime.setDangerouslyCompileStyles.bind(runtime),
};

// We add this to a context so its overridable in tests
export const StoreContext = createContext(runtime);
