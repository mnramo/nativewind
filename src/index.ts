export * from "./provider";
export * from "./use-tailwind";
export * from "./styled";
export * from "./styled-component";
export { useColorScheme } from "./context/color-scheme";
export { useComponent, ComponentContext } from "./context/component";
export { useDeviceMedia } from "./context/device-media";
export { usePlatform } from "./context/platform";
export { useStyleSheet } from "./context/style-sheet";
