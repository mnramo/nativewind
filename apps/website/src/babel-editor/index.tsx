import React, { useMemo, useState } from "react";

import debounce from "lodash.debounce";
import CodeMirror from "@uiw/react-codemirror";
import { dracula } from "@uiw/codemirror-theme-dracula";
import { eclipse } from "@uiw/codemirror-theme-eclipse";
import { javascript } from "@codemirror/lang-javascript";
import { useColorMode } from "@docusaurus/theme-common";

const initialText = `import { Text } from "react-native";

export function MyFancyButton(props) {
  return (
    <Pressable className="component bg-violet-500 hover:bg-violet-600 active:bg-violet-700">
      <Text className="font-bold component-active:font-extrabold" {...props} />;
    </Pressable>
  );
}`;

const initialCompiledText = `import { NativeWindStyleSheet } from "nativewind";
import { StyledComponent } from "nativewind";

/* Example how your code will look */
import { Text } from "react-native";
export function MyFancyButton(props) {
  return <StyledComponent className="component bg-violet-500 hover:bg-violet-600 active:bg-violet-700" component={Pressable}>
      <StyledComponent className="font-bold component-active:font-extrabold" {...props} component={Text} />;
    </StyledComponent>;
}
NativeWindStyleSheet.create({
  styles: {
    "bg-violet-500": {
      "backgroundColor": "#8b5cf6"
    },
    "font-bold": {
      "fontWeight": "700"
    },
    "hover:bg-violet-600": {
      "backgroundColor": "#7c3aed"
    },
    "active:bg-violet-700": {
      "backgroundColor": "#6d28d9"
    }
  },
  masks: {
    "hover:bg-violet-600": 1,
    "active:bg-violet-700": 2
  }
});`;

export default function BabelEditor() {
  const { colorMode } = useColorMode();
  const [compiledText, setCompiledText] = useState(initialCompiledText);
  const update = useMemo(
    () =>
      debounce((code: string) => {
        fetch(
          `https://nativewind-demo-compiler.vercel.app/api/transform?a=${encodeURIComponent(
            code
          )}`
        )
          .then((response) => response.json())
          .then(({ body }) => {
            setCompiledText(body);
          });
      }, 1000),
    [setCompiledText]
  );

  return (
    <div className="babel-editor">
      <h4>Input</h4>
      <CodeMirror
        value={initialText}
        height="300px"
        extensions={[javascript({ jsx: true })]}
        theme={colorMode === "dark" ? dracula : eclipse}
        onChange={update}
      />
      <h4>Output</h4>
      <CodeMirror
        value={compiledText}
        editable={false}
        height="600px"
        extensions={[javascript({ jsx: true })]}
        theme={colorMode === "dark" ? dracula : eclipse}
      />
    </div>
  );
}
