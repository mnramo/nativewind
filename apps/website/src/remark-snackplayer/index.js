/* eslint-disable -- disable all rules */
/**
 * Sections of the source code are licensed under the MIT license found in the
 * LICENSE file in the directory of this source tree. Copyright (c) Facebook, Inc. and its affiliates.
 */

"use strict";
const visit = require("unist-util-visit-parents");
const u = require("unist-builder");
const dedent = require("dedent");

const parseParams = (paramString = "") => {
  const params = Object.fromEntries(new URLSearchParams(paramString));

  if (!params.platform) {
    params.platform = "web";
  }

  return params;
};

const processNode = (node, parent) => {
  return new Promise(async (resolve, reject) => {
    try {
      const params = parseParams(node.meta);

      // Gather necessary Params
      const name = params.name ? decodeURIComponent(params.name) : "Example";
      const description = params.description
        ? decodeURIComponent(params.description)
        : "Example usage";
      const sampleCode = node.value;
      const code = `import React from 'react';
import { withExpoSnack } from 'nativewind';

${sampleCode}

// This demo is using a external compiler that will only work in Expo Snacks.
// You may see flashes of unstyled content, this will not occur under normal use!
// Please see the documentation to setup your application
export default withExpoSnack(App);
`;

      const platform = params.platform || "web";
      const supportedPlatforms = params.supportedPlatforms || "ios,android,web";
      const theme = params.theme || "light";
      const preview = params.preview || "true";
      const loading = params.loading || "lazy";
      const dependencies =
        params.dependencies || "react,react-native,nativewind@latest";

      // Generate Node for SnackPlayer
      // See https://github.com/expo/snack/blob/main/docs/embedding-snacks.md
      const snackPlayerDiv = u("html", {
        value: dedent`
          <div
            class="snack-player"
            data-snack-name="${name}"
            data-snack-description="${description}"
            data-snack-dependencies="${dependencies}"
            data-snack-platform="${platform}"
            data-snack-supported-platforms="${supportedPlatforms}"
            data-snack-theme="${theme}"
            data-snack-preview="${preview}"
            data-snack-loading="${loading}"
            data-snack-sdkversion="45.0.0"
            data-snack-code="${encodeURIComponent(code)}"
          ></div>
          `,
      });

      // Replace code block with SnackPlayer Node
      const index = parent[0].children.indexOf(node);
      parent[0].children.splice(index, 1, snackPlayerDiv);
    } catch (e) {
      return reject(e);
    }
    resolve();
  });
};

const SnackPlayer = () => {
  return (tree) =>
    new Promise(async (resolve, reject) => {
      const nodesToProcess = [];
      // Parse all CodeBlocks
      visit(tree, "code", (node, parent) => {
        // Add SnackPlayer CodeBlocks to processing queue
        if (node.lang == "SnackPlayer") {
          nodesToProcess.push(processNode(node, parent));
        }
      });

      // Wait for all promises to be resolved
      Promise.all(nodesToProcess).then(resolve()).catch(reject());
    });
};

module.exports = SnackPlayer;
