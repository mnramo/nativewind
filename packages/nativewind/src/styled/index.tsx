import {
  createElement,
  ReactNode,
  ComponentType,
  forwardRef,
  RefAttributes,
  ForwardedRef,
  ClassAttributes,
  ForwardRefExoticComponent,
  PropsWithoutRef,
  useContext,
} from "react";
import { StyledProps, StyledPropsWithKeys } from "../utils/styled";
import { InteractionProps, useInteraction } from "./use-interaction";
import { withStyledChildren } from "./with-styled-children";
import { withStyledProps } from "./with-styled-props";
import { useTailwind } from "./use-tailwind";
import { StyleProp } from "react-native";
import { StoreContext } from "../style-sheet";
import { GroupContext, IsolateGroupContext } from "./group-context";
import { useComponentState } from "./use-component-state";
import { GROUP, GROUP_ISO, matchesMask } from "../utils/selector";

export interface StyledOptions<P> {
  props?: Array<keyof P & string>;
  spreadProps?: Array<keyof P & string>;
  classProps?: Array<keyof P & string>;
  baseClassName?: string;
}

type ForwardRef<T, P> = ForwardRefExoticComponent<
  PropsWithoutRef<P> & RefAttributes<T>
>;

type InferRef<T> = T extends RefAttributes<infer R> | ClassAttributes<infer R>
  ? R
  : unknown;

/**
 * Default
 */
export function styled<T>(
  Component: ComponentType<T>
): ForwardRef<InferRef<T>, StyledProps<T>>;

/**
 * Base className
 */
export function styled<T>(
  Component: ComponentType<T>,
  baseClassName: string
): ForwardRef<InferRef<T>, StyledProps<T>>;

/**
 * Base className w/ options
 */
export function styled<T, K extends keyof T & string>(
  Component: ComponentType<T>,
  baseClassName: string,
  options: StyledOptions<T> & {
    props?: Array<K>;
    spreadProps?: Array<K>;
    classProps?: Array<K>;
  }
): ForwardRef<InferRef<T>, StyledPropsWithKeys<T, K>>;

/**
 * Only options
 */
export function styled<T, K extends keyof T & string>(
  Component: ComponentType<T>,
  options: StyledOptions<T> & {
    props?: Array<K>;
    spreadProps?: Array<K>;
    classProps?: Array<K>;
  }
): ForwardRef<InferRef<T>, StyledPropsWithKeys<T, K>>;

/**
 * Actual implementation
 */
export function styled<
  T extends { style?: StyleProp<unknown>; children?: ReactNode | undefined }
>(
  Component: ComponentType<T>,
  styledBaseClassNameOrOptions?: string | StyledOptions<T>,
  maybeOptions: StyledOptions<T> = {}
) {
  const {
    props: propsToTransform,
    spreadProps,
    classProps,
  } = typeof styledBaseClassNameOrOptions === "object"
    ? styledBaseClassNameOrOptions
    : maybeOptions;

  const baseClassName =
    typeof styledBaseClassNameOrOptions === "string"
      ? styledBaseClassNameOrOptions
      : maybeOptions?.baseClassName;

  function Styled(
    {
      className: propClassName = "",
      tw: twClassName,
      style: inlineStyles,
      children: componentChildren,
      ...componentProps
    }: StyledProps<T>,
    ref: ForwardedRef<unknown>
  ) {
    const store = useContext(StoreContext);
    const groupContext = useContext(GroupContext);
    const isolateGroupContext = useContext(IsolateGroupContext);

    const classNameWithDefaults = baseClassName
      ? `${baseClassName} ${twClassName ?? propClassName}`
      : twClassName ?? propClassName;

    /**
     * Get the hover/focus/active state of this component
     */
    const [componentState, dispatch] = useComponentState();

    /**
     * Resolve the props/classProps/spreadProps options
     */
    const {
      styledProps,
      mask: propsMask,
      className,
    } = withStyledProps({
      className: classNameWithDefaults,
      preprocessed: store.preprocessed,
      componentProps,
      propsToTransform,
      classProps,
      spreadProps,
    });

    /**
     * Resolve the className->style
     */
    const style = useTailwind({
      className,
      inlineStyles,
      ...componentState,
      ...groupContext,
      ...isolateGroupContext,
    });

    const mask = (style.mask || 0) | propsMask;

    /**
     * Determine if we need event handlers for our styles
     */
    const handlers = useInteraction(
      dispatch,
      mask,
      componentProps as InteractionProps
    );

    /**
     * Resolve the child styles
     */
    const children = withStyledChildren({
      componentChildren,
      componentState,
      mask,
      store,
      stylesArray: style,
    });

    const element = createElement(Component, {
      ...componentProps,
      ...handlers,
      ...styledProps,
      style: style.length > 0 ? style : undefined,
      children,
      ref,
    } as unknown as T);

    let returnValue: ReactNode = element;

    if (matchesMask(mask, GROUP)) {
      returnValue = createElement(GroupContext.Provider, {
        children: returnValue,
        value: {
          groupHover: groupContext.groupHover || componentState.hover,
          groupFocus: groupContext.groupFocus || componentState.focus,
          groupActive: groupContext.groupActive || componentState.active,
        },
      });
    }

    if (matchesMask(mask, GROUP_ISO)) {
      returnValue = createElement(IsolateGroupContext.Provider, {
        children: returnValue,
        value: {
          isolateGroupHover: componentState.hover,
          isolateGroupFocus: componentState.focus,
          isolateGroupActive: componentState.active,
        },
      });
    }

    return returnValue;
  }

  if (typeof Component !== "string") {
    Styled.displayName = `NativeWind.${
      Component.displayName || Component.name || "NoName"
    }`;
  }

  return forwardRef(Styled);
}
