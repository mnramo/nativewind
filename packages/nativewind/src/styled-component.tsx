import React, { ComponentProps, ComponentPropsWithRef } from "react";
import { styled, StyledProps } from "./styled";

export type StyledComponentProps<P> = StyledProps<P> & {
  component: React.ComponentType<P>;
};

export const StyledComponent = React.forwardRef(
  ({ component, ...options }, ref) => {
    const Component = React.useMemo(() => styled(component), [component]);
    return (
      <Component
        {...(options as unknown as ComponentProps<typeof Component>)}
        ref={ref as ComponentPropsWithRef<typeof Component>["ref"]}
      />
    );
  }
) as <T, P>(
  props: StyledComponentProps<P> & React.RefAttributes<T>
) => React.ReactElement | null;
