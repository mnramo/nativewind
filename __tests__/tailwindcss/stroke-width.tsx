import { Svg, Circle } from "react-native-svg";
import { render } from "@testing-library/react-native";
import { TestProvider } from "./runner";
import { styled } from "../../src";

const cases: Array<[string, string]> = [
  ["0", "0"],
  ["1", "1"],
  ["2", "2"],
];

const StyledCircle = styled(Circle, { valueProps: ["fill", "stroke"] });

describe("Svg - Stroke Width", () => {
  test.each(cases)("stroke-%s", (unit) => {
    const tree = render(
      <TestProvider css={`stroke-${unit}`}>
        <Svg>
          <StyledCircle stroke={`stroke-${unit}`} />
        </Svg>
      </TestProvider>
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
