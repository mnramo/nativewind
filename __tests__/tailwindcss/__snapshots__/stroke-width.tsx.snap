// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Svg - Stroke Width stroke-0 1`] = `
<RNSVGSvgView
  bbHeight="100%"
  bbWidth="100%"
  focusable={false}
  style={
    Array [
      Object {
        "backgroundColor": "transparent",
        "borderWidth": 0,
      },
      Object {
        "flex": 0,
        "height": "100%",
        "width": "100%",
      },
    ]
  }
>
  <RNSVGGroup>
    <RNSVGCircle
      cx={0}
      cy={0}
      r={0}
    />
  </RNSVGGroup>
</RNSVGSvgView>
`;

exports[`Svg - Stroke Width stroke-1 1`] = `
<RNSVGSvgView
  bbHeight="100%"
  bbWidth="100%"
  focusable={false}
  style={
    Array [
      Object {
        "backgroundColor": "transparent",
        "borderWidth": 0,
      },
      Object {
        "flex": 0,
        "height": "100%",
        "width": "100%",
      },
    ]
  }
>
  <RNSVGGroup>
    <RNSVGCircle
      cx={0}
      cy={0}
      propList={
        Array [
          "strokeWidth",
        ]
      }
      r={0}
      strokeWidth={1}
    />
  </RNSVGGroup>
</RNSVGSvgView>
`;

exports[`Svg - Stroke Width stroke-2 1`] = `
<RNSVGSvgView
  bbHeight="100%"
  bbWidth="100%"
  focusable={false}
  style={
    Array [
      Object {
        "backgroundColor": "transparent",
        "borderWidth": 0,
      },
      Object {
        "flex": 0,
        "height": "100%",
        "width": "100%",
      },
    ]
  }
>
  <RNSVGGroup>
    <RNSVGCircle
      cx={0}
      cy={0}
      propList={
        Array [
          "strokeWidth",
        ]
      }
      r={0}
      strokeWidth={2}
    />
  </RNSVGGroup>
</RNSVGSvgView>
`;
